rclpy/rclpy/executors.py

```diff
@@ -597,37 +640,40 @@ def _wait_for_ready_callbacks(
             if tasks:
                 for task, entity, node in tasks:
                     if (not task.executing() and not task.done() and
-                            (node is None or node in nodes_to_use)):
+                            (node is None or nodes_to_use and node in nodes_to_use)):
                         yielded_work = True
                         yield task, entity, node
                 with self._tasks_lock:
                     # Get rid of any tasks that are done
                     self._tasks = list(filter(lambda t_e_n: not t_e_n[0].done(), self._tasks))
 
             # Gather entities that can be waited on
-            subscriptions: List[Subscription] = []
+            subscriptions: List[Subscription[Any, ]] = []
             guards: List[GuardCondition] = []
             timers: List[Timer] = []
-            clients: List[Client] = []
-            services: List[Service] = []
+            clients: List[Client[Any, Any, Any]] = []
+            services: List[Service[Any, Any, Any]] = []
             waitables: List[Waitable[Any]] = []
-            for node in nodes_to_use:
-                subscriptions.extend(filter(self.can_execute, node.subscriptions))
-                timers.extend(filter(self.can_execute, node.timers))
-                clients.extend(filter(self.can_execute, node.clients))
-                services.extend(filter(self.can_execute, node.services))
-                node_guards = filter(self.can_execute, node.guards)
-                waitables.extend(filter(self.can_execute, node.waitables))
-                # retrigger a guard condition that was triggered but not handled
-                for gc in node_guards:
-                    if gc._executor_triggered:
-                        gc.trigger()
-                    guards.append(gc)
+            if nodes_to_use:
```

Comment:
Nit, if I follow correctly `nodes_to_use` is always a list, so we don't need to check `if nodes_to_use` because the `for` loop will do nothing in the case of an empty list.